core/src/main/scala/dagr/core/tasksystem/ScatterGatherFramework.scala

```diff
@@ -0,0 +1,116 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package dagr.core.tasksystem
+
+////////////////////////////////////////////////////////////////////////////////
+// Traits and methods that should be implemented to run scatter gather.
+////////////////////////////////////////////////////////////////////////////////
+
+/** The scatter gather framework.
+  *
+  * There are three distinct steps for scatter gather:
+  *   (1) A single input is split into multiple sub-inputs.
+  *   (2) Each sub-input is transformed into an output.
+  *   (3) Each output is combined to produce a final output.  T
+  * A single task (see class [[SplitInputTask]]) is responsible for step (1), a set of scatter tasks (see
+  * [[ScatterTask]]) are responsible for step (2), and a gather task (see [[GatherTask]]) is responsible for step (3).
+  *
+  * The public methods are the minimum that need to be implemented to use the scatter gather framework.
+  *
+  * Override the protected methods to customize how scatter or gather tasks are generated, for example to perform
+  * recursive merging (ala merge-sort).  Otherwise, mix in traits that do this for you.
+  *
+  * Be sure to mix-in the appropriate trait in the task returned by each method.
+  *
+  * @tparam Input        the input type to the scatter gather.
```

Comment:
I really like `Domain` for this.  I think part of the reason I found this confusing the first time is that for scattering there are really two inputs:
1. The actual data (e.g. a BAM file, a VCF, some other large datasets)
2. The domain in which you are partitioning/scattering the data, e.g. genomic intervals

Since the pattern tends to be applied to indexed or otherwise sub-accessible data, the partitioning is applied to the domain and the scattering is performed by the scatter tasks.

Using a name like `Input` makes me thing we're talking about the bulky data (like BAM or VCF) and not the domain in which the partitioning is done.